packages/back-end/src/models/GeneratedHypothesis.ts

```diff
@@ -0,0 +1,264 @@
+import { GeneratedHypothesisInterface } from "@back-end/types/generated-hypothesis";
+import { ExperimentInterface } from "@back-end/types/experiment";
+import { createExperiment, getExperimentById } from "./ExperimentModel";
+import { upsertWatch } from "./WatchModel";
+import { createVisualChangeset } from "./VisualChangesetModel";
+import { createFeature } from "./FeatureModel";
+
+type GeneratedHypothesisDocument = mongoose.Document &
+  GeneratedHypothesisInterface;
+
+const generatedHypothesisSchema = new mongoose.Schema({
+  id: String,
+  uuid: String,
+  createdAt: Date,
+  organization: String,
+  url: String,
+  hypothesis: String,
+  experiment: {
+    type: String,
+    index: true,
+    require: false,
+  },
+});
+
+generatedHypothesisSchema.index({ uuid: 1 }, { unique: true });
+
+const GeneratedHypothesisModel = mongoose.model<GeneratedHypothesisDocument>(
+  "GeneratedHypothesis",
+  generatedHypothesisSchema
+);
+
+const toInterface = (
+  doc: GeneratedHypothesisDocument
+): GeneratedHypothesisInterface =>
+  omit(doc.toJSON<GeneratedHypothesisDocument>(), ["__v", "_id"]);
+
+export const findOrCreateGeneratedHypothesis = async (
+  context: ReqContext,
+  uuid: string
+): Promise<GeneratedHypothesisInterface> => {
+  const { org, userId } = context;
+  const existing = await GeneratedHypothesisModel.findOne({
+    uuid,
+  });
+  if (existing) return toInterface(existing);
+  if (!process.env.SUPABASE_URL || !process.env.SUPABASE_SERVICE_KEY)
+    throw new Error("Supabase keys missing");
+
+  const supabase = createClient(
+    process.env.SUPABASE_URL,
+    process.env.SUPABASE_SERVICE_KEY
+  );
+
+  const { data: siteHypothesis, error: siteHypErr } = await supabase
+    .from("sites_hypotheses")
+    .select("*, sites(url), sites_dom_mutations(sdk_payload)")
+    .eq("uuid", uuid)
+    .single();
+
+  if (siteHypErr) throw new Error(siteHypErr.message);
+
+  const {
+    hypothesis: { slug, oneLineSummary, hypothesis, control, variant },
+    sites: { url },
+    sites_dom_mutations,
+  } = siteHypothesis;
+  const { sdk_payload } = sites_dom_mutations || {};
+
+  const experimentToCreate: Pick<
+    ExperimentInterface,
+    | "name"
+    | "owner"
+    | "description"
+    | "hypothesis"
+    | "hashAttribute"
+    | "status"
+    | "trackingKey"
+    | "variations"
+    | "phases"
+  > = {
+    name: slug,
+    owner: userId,
+    description: oneLineSummary,
+    hypothesis,
+    hashAttribute: "id",
+    status: "draft",
+    trackingKey: slug,
+    variations: [
+      {
+        name: "Control",
+        description: control,
+        key: "0",
+        screenshots: [],
+        id: uniqid("var_"),
+      },
+      {
+        name: `Variation 1`,
+        description: variant,
+        key: "1",
+        screenshots: [],
+        id: uniqid("var_"),
+      },
+    ],
+    phases: [
+      {
+        coverage: 1,
+        dateStarted: new Date(),
+        dateEnded: new Date(),
```

Comment:
are we supposed to set the end date?

Reply:
I wasn't sure - we create experiments elsewhere in the app like this

Reply:
I removed it - seems to work fine